WCH MCU reorganization

recalci opened the pull request:
According to the official Series Product Naming Rules in the [ch32m030 datasheet.](https://github.com/openwch/ch32m030/tree/main/Datasheet)
![wch_naming_rule](https://github.com/user-attachments/assets/b69a7558-c449-4769-a660-4631c05cf0dc)
WCH ch32 mcus can be categorized into the following series: ch32v0, ch32v1, ch32v2, ch32v3, ch32x0, ch32l1, ch32m0.

![wch_category](https://github.com/user-attachments/assets/37415a35-4584-4330-878e-ddf1b4fb8833)
Additionally, based on the category on the [WCH website in Chinese](https://www.wch.cn/products/productsCenter?categoryId=5), some models that do not start with "ch32" still use the same Qingke series core.

To simplify the process of adding support for other wch mcus, I propose the following organization:
# dts/riscv/wch
- **ch32v0**
    - ch32v003
- **ch32x0**
    - ch32x035
- **ch64**
    - ch641
    - ch643

# soc/wch/ch32v
- **qingke-v2a**
    - ch641
    - ch32v003
- **qingke-v4c**
    - ch32x035
    - ch643

recalci committed:
dts: wch: move ch32v00x.dtsi to ch32v0/ch32v003.dtsi

The CH32V003 belongs to the CH32V0 series. To improve code organization
and maintainability, all related files should be grouped together in a
dedicated subdirectory.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>

recalci committed:
soc: wch: reorganize series directories by core

Place ch32v003 under the qingke_v2a series.
Place qingke series under the ch32v family.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>

recalci committed:
dts: wch: add all ch32v003 packages

Add support for all ch32v003 packages.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>

recalci committed:
soc: wch: add generic vector table support

Add `VECTOR_TABLE_SIZE` Kconfig option to define the number of interrupt
and exception vectors based on the actual hardware specification.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>

fkokosinski commented on dts/riscv/wch/qingke-v2.dtsi:
```diff
@@ -0,0 +1,45 @@
+			#address-cells = <0>;
+			#interrupt-cells = <1>;
+			interrupt-controller;
+			reg = <0xe000e000 16>;
```
```suggestion
			reg = <0xe000e000 0x10>;
```

We usually use hex in reg sizes, so let's keep it consistent here as well

recalci replied:
Done. Thanks.

fkokosinski commented on dts/riscv/wch/qingke-v2.dtsi:
```diff
@@ -0,0 +1,45 @@
+
+		systick: systimer@e000f000 {
+			compatible = "wch,systick";
+			reg = <0xe000f000 16>;
```
```suggestion
			reg = <0xe000f000 0x10>;
```

recalci replied:
Done.

fkokosinski commented on soc/wch/ch32v/Kconfig:
```diff
@@ -0,0 +1,22 @@
+	int "Number of Interrupt and Exception Vectors"
+	range 20 256
+	help
+		This option defines the total number of interrupt and exception
+		vectors in the vector table.
```
```suggestion
	  This option defines the total number of interrupt and exception
	  vectors in the vector table.
```

Help string should be indented with 2 spaces

recalci replied:
Done. Thank you.

fkokosinski commented on dts/riscv/wch/ch32v0/ch32v003a4m.dtsi:
```diff
@@ -0,0 +1,19 @@
+
+#include <wch/ch32v0/ch32v003.dtsi>
+
+&gpioa {
+	gpio-reserved-ranges = <0 1>, <3 5>;
+};
```
Looks like this node stays the same in all dtsis; it probably should be moved to the common dtsi

recalci replied:
Done.

ycsin commented on soc/wch/ch32v/qingke_v2a/vector.S:
```diff
@@ -15,7 +15,7 @@ GTEXT(__initialize)
 	.option norvc
 	j       __start
-	.rept 38
+	.rept   38
```
this line change should be in another commit

recalci replied:
Sorry for my late reply. Done.

ycsin commented on soc/wch/ch32v/Kconfig:
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2024 Michael Hope
```
you should probably use/append your name here

recalci replied:
Since I just split the file into two, I appended my name.

ycsin commented on soc/wch/ch32v/Kconfig.defconfig:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2024 Michael Hope
```
ditto

recalci replied:
done

ycsin commented on soc/wch/ch32v/Kconfig.soc:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2024 Michael Hope
```
ditto

recalci replied:
done

ycsin commented on soc/wch/ch32v/qingke_v2a/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2024 Michael Hope
```
ditto

recalci replied:
done

fkokosinski commented on dts/riscv/wch/qingke-v2.dtsi:
```diff
@@ -0,0 +1,45 @@
+
+		cpu0: cpu@0 {
+			device_type = "cpu";
+			compatible = "wch,qingke-v2";
```
`wch,qingke-v2.yaml` binding should include `ricv,cpus.yaml`, so that the required `riscv,isa` field is enforced for the cpu node

recalci replied:
Thanks for the review! I addressed your comment in a new commit.

recalci committed:
dts: riscv: include riscv,cpus.yaml in qingke-v2

This commit updates the qingke-v2 binding to include `riscv,cpus.yaml`
instead of `cpu.yaml`.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>
